Source/WebCore/page/ShadowRealmGlobalScope.idl

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright (C) 2021 Igalia S.L.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+
+[
+    Exposed=ShadowRealm,
+    // JSCustomMarkFunction,
+    JSLegacyParent=JSShadowRealmGlobalScopeBase,
+    Global=ShadowRealm,
+    IsImmutablePrototypeExoticObject,
+    IsImmutablePrototypeExoticObjectOnPrototype,
```

Comment:
So, in the spec, the global object is just an ordinary object with the prototype %ObjectPrototype%.

We may need a way to encode this in the idl codegeneration. There are some things which look like they do something similar, but they aren't exactly what we want --- for instance [LegacyNoInterfaceObject] makes the interface an ordinary object, like a namespace -- this would be fine, but we don't want it to be a namespace, it needs to be the global object itself. So I'm not sure that this works in conjunction with [Global], or if it affects the way it exposes itself on the global object, or what. It's worth a try though.

I don't think we need IsImmutablePrototype... etc --- %Object.prototype% is an immutable prototype object, and that is the prototype of the global object, so that should be all we need. I think these extended attributes are mainly used to tell the codegenerator to set new interface prototypes as immutable prototypes, which doesn't really apply here since we don't have a real "interface" in the first place.

It's a bit weird and maybe hard to explain, but the model of WebIDL in general is like "everything is an instance of an interface, the interface is a function with a prototype", but for us, there is no interface, there is no instance of an interface, it's just an ordinary JS object. If that makes sense

Reply:
like in the usual model, Node is an interface, which is exposed on the global object as a function, so you can do `thing instance of Node` or whatever. (but you can't actually call them directly, normally) -- we don't actually have an interface, there is no ShadowRealmGlobalObject constructor or anything anywhere, it's just an ordinary JS object, not an instance of any interface.

I think we may need to modify the IDL processor to handle this properly, because the existing solutions for namespace objects don't apply directly to this use case